lib/ocamlorg_web/graphql.mli

```diff
@@ -5,41 +5,85 @@ type package_info =
   ; constraints : string option
   }
 
-type packages_result =
+type packages_success =
   { total_packages : int
   ; packages : Package.t list
   }
 
-val starts_with : string -> string -> bool
-(** [starts_with] function compares two strings and returns true if the first
-    string is equal to the beginning of the second string up to capitalization *)
+type packages_result = (packages_success, [ `Msg of string ]) result
 
-val is_package : string -> string -> bool
-(** [is_package] function compares two strings and returns true if they are
-    equal up to capitalization *)
+type package_success = { package : Package.t }
 
-val get_packages_result
-  :  int
+type package_result = (package_success, [ `Msg of string ]) result
+
+type params_validity =
+  | Valid_params
+  | Wrong_limit
+  | Wrong_offset
+
+val get_info : (Package.Name.t * string option) list -> package_info list
+(** [get_info] function loops through the list part of Package.Info such as
+    Package.Info.dependencies and returns a list of name and constraints record
+    Package.Info record *)
+
+(* val is_in_range : 'a -> 'a -> 'a -> bool *)
+val is_in_range : 'a -> 'a option -> 'a option -> bool
+(** [is_in_range] function takes a [current_version] and an optional
+    [from_version] and [upto_version] and checks that current_version is between
+    the range of from_version and upto_version *)
+
+val is_valid_params : int -> int -> int -> params_validity
+(** [is_valid_params] function is a sub function of [all_packages_result] that
+    takes [offset] [limit] [all_packages] and retruns an error message if limit
+    or ossfest is less than 0 or gretaer than [all_packages]*)
+
+val packages_list
+  :  string option
   -> int
   -> int
-  -> string option
   -> Package.t list
-  -> packages_result
-(** [get_packages_result total_packages offset limit startswith packages]
-    returns a list of the latest version of all packages with [limit] and
-    [offset] options which is used to cut out some parts of the
-    [packages_result] list. It also takes a [startswith] option which will
-    search the list of packages looking for the package with a name that begins
-    with [startswith] *)
-
-val get_single_package : Package.t list -> string -> Package.t option
-(** [get_single_package packages name] will search the list of packages looking
-    for the package with a name that matches [name], [None] is returned if no
-    package can be found. *)
+  -> Package.state
+  -> Package.t list
+(** [packages_list] function is a sub function of [all_packages_result] that
```

Comment:
I think this documentation comment could be improved slightly as it doesn't explain what `contains` does, namely it first uses `Packages.search` to get a list of packages then applies the offset and limit. 

It might also be a little more ergonomic to make the `contains` an optional, labelled parameter so users don't have to explicitly pass `None` when they don't want it i.e.

```ocaml
val packages_list
  :  ?contains:string
  -> int
  -> int
  -> Package.t list
  -> Package.state
  -> Package.t list
```

Reply:
I have this now: 
```
   let packages_list ?(contains = None) offset limit all_packages t =
  match contains with
  | None ->
    List.filteri (fun i _ -> offset <= i && i < offset + limit) all_packages
  | Some letters ->
    List.filteri
      (fun i _ -> offset <= i && i < offset + limit)
      (Package.search_package t letters)
 ```

Reply:
Yep this is great, you shouldn't need to explicitly write `?(contains = None)` it is the same as `?contains`.

Reply:
Yea, but it throws an error here 
```
     ~resolve:(fun _ () contains offset limit ->
            Lwt.return (all_packages_result contains offset limit t))
```

Reply:
You probably need: 

```ocaml
     ~resolve:(fun _ () contains offset limit ->
            Lwt.return (all_packages_result ~contains offset limit t))
```

Also just as a minor point, "throws an error" is usually used for your code running and an exception being thrown for example `10 / 0` throws (or raises) the `Division_by_zero` exception. Here, your code doesn't type check (or at least compile, they're basically the same in OCaml for a majority of the cases). The only reason I add this is I started looking for something in your code that raises an exception but then realised the problem :))